sched/sched/sched_note.c

```diff
@@ -405,6 +405,10 @@ void sched_note_start(FAR struct tcb_s *tcb)
   strncpy(note.nsa_name, tcb->name, CONFIG_TASK_NAME_SIZE + 1);
 
   length = SIZEOF_NOTE_START(namelen + 1);
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_RECORD_TASKNAME
+  sched_note_add_taskname(tcb->pid, tcb->name);
```

Comment:
Why need add new api sched_note_add_taskname? we can check NOTE_START in sched_note_add and then extract and save the name.

Reply:
We might have no need to call sched_note_add_taskname() in sched_note_start() because trace dump already retrieves the task name in NOTE_START.
But it still needs in sched_note_stop/suspend/resume() because we have no way to get the name of the task which is already exists before the trace starts. It is the reason I have added new API.

As another solution, it can be solved by adding char name[] member in struct note_stop/suspend/resume_s like note_start_s. But it increases the size of trace data because each context switch records the task name and I think it is not a good idea.

Reply:
> We might have no need to call sched_note_add_taskname() in sched_note_start() because trace dump already retrieves the task name in NOTE_START.

How about we don't clear the name buffer in noteram_buffer_clear? but
1. add task name to the buffer in NOTE_START
2. remove task name from the buffer in NOTE_STOP

> But it still needs in sched_note_stop/suspend/resume() because we have no way to get the name of the task which is already exists before the trace starts. It is the reason I have added new API.
>

it isn't an issue if we don't clear the name buffer?

> As another solution, it can be solved by adding char name[] member in struct note_stop/suspend/resume_s like note_start_s. But it increases the size of trace data because each context switch records the task name and I think it is not a good idea.

Reply:
> 1. add task name to the buffer in NOTE_START
> 2. remove task name from the buffer in NOTE_STOP

Then, the buffer always keeps the name of currently existing tasks.
On the other hand, the trace data is the history of the task activities and it may contain the task which is already terminated when dumping the trace data.
The name buffer must have the name of tasks which made any activities while getting the trace.

Reply:
But how to handle this case:
1.Task 1 create with pid
2.Task 1 exist
3.Task 2 create with the same pid
4.Other trace overwrite the first trace
5.Userspace read out the trace data, how can he get the first task name?

Reply:
To resolve the above issue, one method is:
1.Save task name in NOTE_START
2.Save task name in NOTE_STOP
Then we have these combination:
1.No data lose get task name from NOTE_START
2.NOTE_START lose, but task doesn't terminate, get the task name from pid
3.NOTE_START lose and task terminate, but NOTE_STOP not lose, get task name from NOTE_STOP
4.Both NOTE_START and NOTE_STOP lose, the trace buffer shouldn't have any info related to this task.

Reply:
> 3.Task 2 create with the same pid

I think we don't have a way to do this intentionally.
Kernel tries to assign an unique pid for new task and it takes a long time to assign the same pid which is used in past.

Reply:
If so, we don't need add the new API between sched and note driver:

1. Save the task name in NOTE_START
2. Remove the task name when user read the NOTE_STOP from buffer

Reply:
Currently, noteram driver doesn't treat the contents of note struct. Only nc_length is used to find the length of each notes.
If we add parsing incoming note struct in the driver, we can avoid to add the new API between sched and noteram.
I got for it.

> 1. Save the task name in NOTE_START
> 2. Remove the task name when user read the NOTE_STOP from buffer
> Do you think it's workable?

There are a few concerns.

1. While the task trace is stopped, we cannot capture task creation/deletion by notes.
   For example, if we turn off the tracing on boot, we never able to get the name of idle task and init task.
2. Reading NOTE_STOP doesn't mean the name of the task become needless.
  Because the trace buffer can be read any times until cleared explicitly.

So how about the following proposals?
- Leave sched_note.c as it is. Don't add the new API.
- noteram_driver checks nc_pid[] member of incoming note struct passed by sched_note_add().
- Check whether the pid exists in the name buffer. If not, get the task name and add into the buffer (same as sched_note_add_taskname() of this PR).
- Provide the same interface for userland (NOTERAM_GETTASKNAME ioctl) as this PR.

Reply:
@xiaoxiang781216 I have force pushed new commits.
Now all updates are in noteram driver. It just records the name of the task which calls sched_note_add().

Reply:
> 
> There are a few concerns.
> 
> 1. While the task trace is stopped, we cannot capture task creation/deletion by notes.
>    For example, if we turn off the tracing on boot, we never able to get the name of idle task and init task.
> 2. Reading NOTE_STOP doesn't mean the name of the task become needless.
>    Because the trace buffer can be read any times until cleared explicitly.

Sorry I use the wrong term here, The reading NOTE_STOP really mean that the coming data overwrite the trace buffer
where contain NOTE_STOP.

> 
> So how about the following proposals?
> 
> * Leave sched_note.c as it is. Don't add the new API.
> * noteram_driver checks nc_pid[] member of incoming note struct passed by sched_note_add().
> * Check whether the pid exists in the name buffer. If not, get the task name and add into the buffer (same as sched_note_add_taskname() of this PR).
> * Provide the same interface for userland (NOTERAM_GETTASKNAME ioctl) as this PR.

But one issue still exist: since the name buffer stop to recoard the new task name when the buffer is full, but user can select to stop or overwrite the trace buffer, then the task name can't recovery in a long run.

Reply:
> Sorry I use the wrong term here, The reading NOTE_STOP really mean that the coming data overwrite the trace buffer
> where contain NOTE_STOP.

I got. I have supported this case.